librz/core/rop.c

```diff
@@ -0,0 +1,1344 @@
+				hit->addr, asm_op_hex, rz_asm_op_get_asm(asmop), comment);
+		} else {
+			rz_cons_printf("  0x%08" PFMT64x " %18s  %s\n",
+				hit->addr, asm_op_hex, rz_asm_op_get_asm(asmop));
+		}
+	}
+	free(asm_op_hex);
+	rz_analysis_op_fini(&aop);
+	rz_asm_op_free(asmop);
+	return 0;
+}
+
+static int rz_rop_print_json_mode(const RzCore *core, const RzCoreAsmHit *hit, unsigned int *size, PJ *pj) {
+	RzAnalysisOp aop = RZ_EMPTY;
+	RzAsmOp *asmop = rz_asm_op_new();
+	if (!asmop) {
+		return -1;
+	}
+	if (rz_rop_process_asm_op(core, hit, asmop, &aop, size, NULL, NULL) != 0) {
+		rz_asm_op_free(asmop);
+		return -1;
+	}
+
+	pj_o(pj);
+	pj_kn(pj, "offset", hit->addr);
+	pj_ki(pj, "size", hit->len);
+	pj_ks(pj, "opcode", rz_asm_op_get_asm(asmop));
+	pj_ks(pj, "type", rz_analysis_optype_to_string(aop.type));
+	pj_end(pj);
+
+	rz_analysis_op_fini(&aop);
+	rz_asm_op_free(asmop);
+	return 0;
+}
+
+RZ_IPI void rz_core_rop_reg_info_free(RzRopRegInfo *reg_info) {
+	if (!reg_info) {
+		return;
+	}
+	free(reg_info->name);
+	free(reg_info);
+}
+
+RZ_IPI RzRopRegInfo *rz_core_rop_reg_info_new(const RzCore *core, const RzILEvent *evt, const ut64 init_val, const ut64 new_val) {
+	RzRopRegInfo *reg_info = RZ_NEW0(RzRopRegInfo);
+	const char *name = NULL;
+	if (evt->type == RZ_IL_EVENT_VAR_READ) {
+		reg_info->is_var_read = true;
+		name = evt->data.var_read.variable;
+	} else if (evt->type == RZ_IL_EVENT_VAR_WRITE) {
+		reg_info->is_var_write = true;
+		name = evt->data.var_write.variable;
+	}
+	if (!reg_info) {
+		return NULL;
+	}
+	const RzList *head = rz_reg_get_list(core->analysis->reg, RZ_REG_TYPE_GPR);
+	if (!head) {
+		free(reg_info);
+		return NULL;
+	}
+	RzListIter *iter_dst;
+	RzRegItem *item_dst;
+	rz_list_foreach (head, iter_dst, item_dst) {
+		if (RZ_STR_EQ(name, item_dst->name) && item_dst->type == RZ_REG_TYPE_GPR) {
+			reg_info->name = strdup(name);
+			break;
+		}
+	}
+
+	if (!reg_info->name) {
+		free(reg_info);
+		return NULL;
+	}
+	reg_info->init_val = init_val;
+	reg_info->new_val = new_val;
+	return reg_info;
+}
+
+RZ_API RzRopGadgetInfo *rz_core_rop_gadget_info_new(const ut64 address) {
+	RzRopGadgetInfo *gadget_info = RZ_NEW0(RzRopGadgetInfo);
+	if (!gadget_info) {
+		return NULL;
+	}
+	gadget_info->address = address;
+	gadget_info->stack_change = 0LL;
+	gadget_info->curr_pc_val = address;
+	gadget_info->is_pc_write = false;
+	gadget_info->is_syscall = false;
+	gadget_info->modified_registers = rz_pvector_new((RzPVectorFree)rz_core_rop_reg_info_free);
+	gadget_info->dependencies = rz_list_newf((RzListFree)rz_core_rop_reg_info_free);
+	return gadget_info;
+}
+
+RZ_API void rz_core_rop_gadget_info_free(RzRopGadgetInfo *gadget_info) {
+	if (!gadget_info) {
+		return;
+	}
+	rz_pvector_free(gadget_info->modified_registers);
+	rz_list_free(gadget_info->dependencies);
+	free(gadget_info);
+}
+
+RZ_API void rz_core_rop_gadget_info_add_register(RzRopGadgetInfo *gadget_info, RzRopRegInfo *reg_info, const bool is_dependency) {
```

Comment:
For every `RZ_API` that you add or change, except maybe obvious basic `*_free()` and `*_new()` functions, please add a Doxygen documentation. Moreover, for the external API it's inmportant to provide hints like `RZ_BORROW`, `RZ_OUT`, `RZ_NULLABLE`, `RZ_NONNULL` to the corresponding parameters, and for those marked with `RZ_NONNULL` add `rz_return_*()` asserts at the beginning of the function/